**Transformers From Scratch (Part 1): The Bridge Between Language and Maths**





The Story So Far...
We live in the era of Artificial Intelligence, dominated by massive models that can write code, compose poetry, and summarize textbooks in seconds. But beneath the polished interfaces of ChatGPT, Claude, and Llama lies a single, revolutionary architecture introduced by Google in 2017: The Transformer.

For many developers and data scientists, the Transformer remains a "black box." We know how to import it from Hugging Face in three lines of code, but what is actually happening under the hood? How does raw text turn into "understanding"?

**The Goal of This Series**

This series - Transformers From Scratch - is about breaking the glass. We are going to build the Transformer architecture completely from scratch using PyTorch. No black boxes, no hidden layers just pure code and intuition.

By the end of this journey, you will have a deep, mathematical, and programmatic understanding of how these engines work.

Our Roadmap will Start With:



*   Part 1: (We are Here) Tokenization & Input Embeddings
*   Part 2: Positional & Rotary Embeddings (RoPE)

*   Part 3: Self-Attention & Multi-Head Attention
*   Part 4: The Feed-Forward Network & Layer Normalization
*   Part 5: Assembling the Full Encoder Architecture

*   Part 6: Similar Journey for Decoder

Let's begin at step zero: getting our data ready.

-----------------------------------

**The Blank Canvas: Why Do We Need Tokenization?**

Imagine handing a dense novel to a calculator and asking it to find the main theme. The calculator would stare blankly at you because it only understands numbers, not the letters "A" through "Z."

Machine Learning models, at their core, are just highly complex calculators. They process matrices, perform dot products, and optimize gradients. They have zero concept of what a "word" is.

*Tokenization is the bridge between human language and machine mathematics*.

-It is the process of breaking down raw, unstructured text into smaller, manageable chunks called "tokens," and then mapping each unique token to a specific integer ID. Before our model can learn relationships between words, it must first be able to read them.

*Text $\rightarrow$ Tokens $\rightarrow$ Integer IDs $\rightarrow$ Neural Network*

-------------------------

In [12]:
with open('/content/alice_in_wonderland.txt','r',encoding = 'utf-8') as f:
  raw_text = f.read()

print("The Total Number of character is : ",len(raw_text))
print(raw_text[:210])

The Total Number of character is :  148208
TITLE: Alice's Adventures in Wonderland
AUTHOR: Lewis Carroll


= CHAPTER I = 
=( Down the Rabbit-Hole )=

  Alice was beginning to get very tired of sitting by her sister
on the bank, and of having nothing to 


Our goal is to tokenize this 148208 character short story into individual words and special
characters that we can then turn into embeddings for LLM training  

How can we split the text into tokens best with ? - regex library of Python

In [13]:
import re
text = "Hello World, I am Ayush and This is the Tokenization From Scratch."
result = re.split(r'(\s)',text)

print(result)

['Hello', ' ', 'World,', ' ', 'I', ' ', 'am', ' ', 'Ayush', ' ', 'and', ' ', 'This', ' ', 'is', ' ', 'the', ' ', 'Tokenization', ' ', 'From', ' ', 'Scratch.']


The result is a list of individual words, whitespaces, and punctuation characters along with white spaces (It is optional to remove white spaces):

In [17]:
result = re.split(r'([.,]|\s)',text)
result = [item for item in result if item.strip()]
print(result)

['Hello', 'World', ',', 'I', 'am', 'Ayush', 'and', 'This', 'is', 'the', 'Tokenization', 'From', 'Scratch', '.']


**The Whitespace Dilemma: To Strip or Not to Strip?**

When developing a simple tokenizer, the decision to include or remove whitespaces is a trade-off between structural precision and computational efficiency. Removing whitespaces is a common choice for introductory projects because it significantly reduces the sequence length and memory requirements, allowing us to focus purely on the relationship between semantic tokens (words and punctuation) without the added noise of "invisible" characters. However, while this approach ensures brevity and simplicity in our initial outputs, it does strip away the model's ability to learn specific formatting or indentation—a critical feature for tasks like code generation. For the purpose of this first stage, we will prioritize simplicity and remove whitespaces to keep our tokenized outputs clean, with the plan to transition to a more sophisticated, whitespace-aware scheme as our architecture grows in complexity.